pkg/route/validation/validation.go

```diff
@@ -342,6 +382,62 @@ func validateWildcardPolicy(host string, policy routev1.WildcardPolicyType, fldP
 	return nil
 }
 
+var (
+	notAllowedHTTPHeaders        = []string{"strict-transport-security", "proxy", "cookie", "set-cookie"}
+	notAllowedHTTPHeaderSet      = sets.NewString(notAllowedHTTPHeaders...)
+	notAllowedHTTPHeadersMessage = fmt.Sprintf("the following headers may not be modified using this API: %v", strings.Join(notAllowedHTTPHeaders, ", "))
+)
+
+// validateHeaders verifies that the given slice of request or response headers
+// is valid using the given regexp.
+func validateHeaders(fldPath *field.Path, headers []routev1.RouteHTTPHeader, valueRegexp *regexp.Regexp, valueErrorMessage string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	headersMap := map[string]struct{}{}
+	for i, header := range headers {
+		idxPath := fldPath.Index(i)
+
+		// Each action must specify a unique header.
+		_, alreadyExists := headersMap[header.Name]
+		if alreadyExists {
+			err := field.Duplicate(idxPath.Child("name"), header.Name)
+			allErrs = append(allErrs, err)
+		}
+		headersMap[header.Name] = struct{}{}
+
+		if len(header.Name) == 0 {
+			err := field.Required(idxPath.Child("name"), "")
+			allErrs = append(allErrs, err)
+		}
+		if len(header.Name) > maxHeaderNameSize {
+			err := field.Invalid(idxPath.Child("name"), header.Name, fmt.Sprintf("name exceeds the maximum length, which is 1024"))
+			allErrs = append(allErrs, err)
+		}
+		if notAllowedHTTPHeaderSet.Has(strings.ToLower(header.Name)) {
+			err := field.Forbidden(idxPath.Child("name"), notAllowedHTTPHeadersMessage)
+			allErrs = append(allErrs, err)
+		}
+
+		if header.Action.Type == routev1.Set && header.Action.Set == nil || header.Action.Type != routev1.Set && header.Action.Set != nil {
+			err := field.Required(idxPath.Child("action", "set"), "set is required when type is Set, and forbidden otherwise")
+			allErrs = append(allErrs, err)
+		}
+		if header.Action.Set != nil {
+			switch valueLength := len(header.Action.Set.Value); {
+			case valueLength == 0:
+				err := field.Required(idxPath.Child("action", "set", "value"), "")
+				allErrs = append(allErrs, err)
+			case valueLength > maxHeaderValueSize:
+				err := field.Invalid(idxPath.Child("action", "set", "value"), header.Action.Set.Value, "value exceeds the maximum length, which is 16384")
+				allErrs = append(allErrs, err)
+			case !valueRegexp.MatchString(header.Action.Set.Value):
+				err := field.Invalid(idxPath.Child("action", "set", "value"), header.Action.Set.Value, valueErrorMessage)
+				allErrs = append(allErrs, err)
+			}
+		}
```

Comment:
There should be  a validation check for `header.Action.Delete`, did it get removed?  If the unit test for `header.Action.Delete` passed without error, then that test needs to be re-examined.

Reply:
Do you have an example test case, or example of the validation you expect to see?  I see these negative test cases for the "Delete" action:

* "Should give an error on attempt to delete the HSTS header."
* "Should give an error on attempt to delete the Proxy header."
* "Should give an error on attempt to delete the Cookie header."
* "Should give an error on attempt to delete the Set-Cookie header."
* "should not allow repetition of a header name".
* "set is forbidden when type is not Set".

And these positive test cases with the "Delete" action:

* "Valid response headers".
* "Valid request headers".
* "Valid request and response headers".

Reply:
we check for presence of header name before so this is not needed from my understanding

Reply:
How about making sure the header.Action.Type is either Delete or Set, and nothing else?

Reply:
added